app/eth2wrap/synthproposer.go

```diff
@@ -0,0 +1,206 @@
+	}
+}
+
+// synthProposerCache caches actual and synthetic proposer duties for the set of public keys.
+//
+// Since only a single validator can be a proposer per slot, we require all
+// validators to calculate the synthetic duties for the whole set.
+type synthProposerCache struct {
+	pubkeys []eth2p0.BLSPubKey
+	// shuffleFunc deterministically shuffles the validator indices for the epoch.
+	shuffleFunc func(eth2p0.Epoch, []eth2p0.ValidatorIndex) []eth2p0.ValidatorIndex
+
+	mu     sync.RWMutex
+	fifo   []eth2p0.Epoch
+	duties map[eth2p0.Epoch][]*eth2v1.ProposerDuty
+	synths map[eth2p0.Epoch]map[eth2p0.Slot]bool
+}
+
+// Duties returns the upstream and synthetic duties for all pubkeys for the provided epoch
+// via a cache.
+func (c *synthProposerCache) Duties(ctx context.Context, eth2Cl synthProposerEth2Provider, epoch eth2p0.Epoch) ([]*eth2v1.ProposerDuty, error) {
+	// Check if cache already populated for this epoch using read lock.
+	c.mu.RLock()
+	duties, ok := c.duties[epoch]
+	c.mu.RUnlock()
+	if ok {
+		return duties, nil
+	}
+
+	// Get slotsPerEpoch and the starting slot of the epoch.
+	slotsPerEpoch, err := eth2Cl.SlotsPerEpoch(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	epochSlot := eth2p0.Slot(epoch) * eth2p0.Slot(slotsPerEpoch)
+
+	// Get active validators for the epoch
+	vals, err := eth2Cl.ValidatorsByPubKey(ctx, fmt.Sprint(epochSlot), c.pubkeys)
+	if err != nil {
+		return nil, err
+	}
+
+	var activeIdxs []eth2p0.ValidatorIndex
+	for _, val := range vals {
+		if !val.Status.IsActive() {
+			continue
+		}
+		activeIdxs = append(activeIdxs, val.Index)
+	}
+
+	// Get actual duties for all validators for the epoch.
+	duties, err = eth2Cl.ProposerDuties(ctx, epoch, activeIdxs)
+	if err != nil {
+		return nil, err
+	}
+
+	// Mark those not requiring synthetic duties.
+	noSynth := make(map[eth2p0.ValidatorIndex]bool)
+	for _, duty := range duties {
+		noSynth[duty.ValidatorIndex] = true
+	}
+
+	// Deterministic synthetic duties for the rest.
+	synthSlots := make(map[eth2p0.Slot]bool)
+	for _, valIdx := range c.shuffleFunc(epoch, activeIdxs) {
+		if noSynth[valIdx] {
+			continue
+		}
+
+		offset := eth2p0.Slot(valIdx) % eth2p0.Slot(slotsPerEpoch)
+		synthSlot := epochSlot + offset
+		if _, ok := synthSlots[synthSlot]; ok {
+			// We already have a synth proposer for this slot.
+			continue
+		}
+
+		synthSlots[synthSlot] = true
+		duties = append(duties, &eth2v1.ProposerDuty{
+			PubKey:         vals[valIdx].Validator.PublicKey,
+			Slot:           synthSlot,
+			ValidatorIndex: valIdx,
+		})
+	}
+
+	// Cache the values for the epoch
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.fifo = append(c.fifo, epoch)
+	c.duties[epoch] = duties
+	c.synths[epoch] = synthSlots
+
+	// Trim the cache
+	if len(c.fifo) > maxCachedEpochs {
+		delete(c.duties, c.fifo[0])
+		delete(c.synths, c.fifo[0])
+		c.fifo = c.fifo[1:]
+	}
+
+	return duties, nil
+}
+
+// IsSynthetic returns true if the slot is a synthetic proposer duty from via a cache.
```

Comment:
```suggestion
// IsSynthetic returns true if a synthetic proposer duty from cache is scheduled in the given slot.
```

Reply:
removed reference to cache since they type is `synthProposerCache`, so cache is implied.